tests/harness.rs

```diff
@@ -0,0 +1,307 @@
+//! and a submit channel so tests can manually control the simulation environment,
+//! ticking along new blocks and asserting on the state of the simulator's results.
+//!
+//! It also manages two Anvil instances to simulate the rollup and host chains,
+//! and provides utility functions for advancing blocks and adding transactions to
+//! the simulator's mempool.
+
+use std::{
+    sync::Arc,
+    time::{Duration, SystemTime, UNIX_EPOCH},
+};
+
+use alloy::{
+    consensus::Header,
+    eips::{BlockId, eip1559::BaseFeeParams},
+    network::{Ethereum, EthereumWallet},
+    node_bindings::Anvil,
+    primitives::{B256, U256},
+    providers::{
+        Provider, ProviderBuilder, RootProvider,
+        ext::AnvilApi,
+        fillers::{BlobGasFiller, SimpleNonceManager},
+        layers::AnvilProvider,
+    },
+    signers::local::PrivateKeySigner,
+};
+use builder::{
+    config::{BuilderConfig, HostProvider},
+    tasks::{
+        block::sim::{SimResult, Simulator},
+        env::{Environment, SimEnv},
+    },
+    test_utils::{new_signed_tx_with_max_fee, setup_logging, setup_test_config},
+};
+use init4_bin_base::utils::calc::SlotCalculator;
+use signet_sim::SimCache;
+use tokio::{
+    sync::{mpsc, watch},
+    task::JoinHandle,
+};
+use trevm::revm::{context::BlockEnv, context_interface::block::BlobExcessGasAndPrice};
+
+// Default test slot duration (seconds)
+const DEFAULT_SLOT_DURATION: u64 = 5; // seconds
+const TEST_TX_MAX_FEE: u128 = 1_000_000_000_000;
+
+pub struct SimulatorTask {
+    sim_env_tx: watch::Sender<Option<SimEnv>>,
+    sim_env_rx: watch::Receiver<Option<SimEnv>>,
+    sim_cache: SimCache,
+}
+
+pub struct TestHarness {
+    /// Builder configuration for the Harness
+    pub config: BuilderConfig,
+    /// Anvil provider made from the Rollup Anvil instance
+    pub rollup: RollupAnvilProvider,
+    /// Anvil provider made from the Host Anvil instance
+    pub host: HostAnvilProvider,
+    /// The Simulator task that is assembled each tick.
+    pub simulator: SimulatorTask,
+    /// Transaction plumbing - Submit
+    pub submit_tx: mpsc::UnboundedSender<SimResult>,
+    /// Transaction plumbing - Receive
+    pub submit_rx: mpsc::UnboundedReceiver<SimResult>,
+    /// Keeps the simulator task alive for the duration of the harness so the
+    /// background task isn't aborted when `start()` returns.
+    pub simulator_handle: Option<JoinHandle<()>>,
+}
+
+type HostAnvilProvider = AnvilProvider<RootProvider<Ethereum>>;
+type RollupAnvilProvider = AnvilProvider<RootProvider<Ethereum>>;
+
+type RollupHeader = Header;
+type HostHeader = Header;
+type Headers = (RollupHeader, HostHeader);
+
+impl TestHarness {
+    /// Create a new harness with a fresh Anvil rollup chain and default test config.
+    pub async fn new() -> eyre::Result<Self> {
+        setup_logging();
+        let mut config = setup_test_config()?;
+        configure_slot_timing(&mut config)?;
```

Comment:
this is an anti-pattern. if we want the config to have a specific value, we should set it up that way. i;m not sure why we want to change this at all. we should be using either the test value in constants, or the pecorino value in constants